### PR TITLE
OSD::Modify OSD Fast-Shutdown to work safely i.e. quiesce all activit…

### PR DESCRIPTION
**Pull request by benhanokh**
OSD::Modify OSD Fast-Shutdown to work safely i.e. quiesce all activities and destage allocations to disk before killing the OSD

1) keep the old (unsafe) fast-shutdown when we are not using NCB (non null-manager())
2) skip service.prepare_to_stop() which can take as much as 10 seconds
3) skip debug options in fast-shutdown
4) set_state(STATE_STOPPING) which will stop accepting new tasks to this OSD
5) clear op_shardedwq queues, this is safe since we didn't started processing them
6) stop timer
7) drain osd_op_tp (no new items will be added)
8) now we can safely call umount which will close_db/bluefs and will destage allocation to disk
9) skip _shutdown_cache() when we are in the middle of a fast-shutdown
10) increase debug level on fast-shutdown
11) add option for bluestore_qfsck_on_mount to force scan on mount for all tests
12) disable fsck-on-umount when running fast-shutdown
13) add an option to increase debug level at fast-shutdown umount()
14) set a time limit to fast-shutdown
15) Bug-Fix BlueStore::pool_statfs don't access db after it was removed
16) Fix error message for qfsck (error was caused by PR https://github.com/ceph/ceph/pull/44563)

Fixes: https://tracker.ceph.com/issues/53266

Signed-off-by: Gabriel Benhanokh <gbenhano@redhat.com>

### CODE REVIEW

File: qa/tasks/ceph.py [hunk 1]
```diff
@@ -1214,6 +1214,40 @@ def first_in_ceph_log(pattern, excludes):
             ),
         )
 
+def osd_scrub_bluestore(ctx, config):
+    """
+    Scrub bluestore allocation after we exit making sure allocation file is consistent with ONodes.
+    OSDs must be down for the command to work!
+    """
+
+    self.log('osd_scrub_bluestore started')
```
Comment (jdurgin) [inline qa/tasks/ceph.py:1223]: this function uses self a bunch, which isn't defined here. It looks like it's based on code from ceph_manager, but as written this won't work.

When you're testing a change like this you need to also specify your branch as the qa suite branch so that these changes to teuthology tasks are picked up too - e.g. use ```--suite-branch safe_shutdown_v2``` when scheduling the test run

Comment (jdurgin) [reply qa/tasks/ceph.py:1223]: I'd suggest keeping this implementation in ceph_manager.py - you could make it a method for one osd,  ```osd_scrub_bluestore(self, osd_id)``` and loop over the osds here in ceph.py. You can get the relevant CephManager instance here via: ```ceph_manager = ctx.managers[config['cluster']]```

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -4287,6 +4304,51 @@ int OSD::shutdown()
...
+    std::lock_guard lock(osd_lock);
+    //
+    // assert in allocator that nothing is being add
+    //
+    store->umount();
+    //store.reset();
+
+    utime_t end_time = ceph_clock_now();
+    // fast-shutdown should complete within 15 seconds
+    ceph_assert(end_time - start_time_func < 15);
```
Comment (jdurgin) [inline src/osd/OSD.cc:4340]: todo: make timeout configurable

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -4287,6 +4304,51 @@ int OSD::shutdown()
     cct->_conf.apply_changes(nullptr);
   }
 
+  if (cct->_conf->osd_fast_shutdown) {
+    // first, stop new task from being taken from op_shardedwq
+    op_shardedwq.stop_for_fast_shutdown();
+
+    // do we still need to drain ???
```
Comment (jdurgin) [inline src/osd/OSD.cc:4311]: not when we just discarded the queue!

Comment (jdurgin) [reply src/osd/OSD.cc:4311]: the requests that are in flight are being processed by osd_op_tp, which you drain below, so they should be handled already. in fact, wq->drain() just calls tp->drain()

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -4287,6 +4304,51 @@ int OSD::shutdown()
...
+    //op_shardedwq.drain();
+
+    utime_t  start_time_timer = ceph_clock_now();
+    // TBD: Is this too early???
```
Comment (jdurgin) [inline src/osd/OSD.cc:4315]: I think this should be ok at this point. May want to stop tick_timer_without_osd_lock as well - it could have generated more work (which is being discarded, but it's still not necessary)

Comment (jdurgin) [inline src/osd/OSD.cc:4315]: should be ok at this point

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -10305,6 +10337,11 @@ int BlueStore::get_numa_node(
   return 0;
 }
 
+void BlueStore::prepare_for_fast_shutdown()
+{
+  m_fast_shutdown = true;
```
Comment (jdurgin) [inline src/os/bluestore/BlueStore.cc:10342]: should we check this in queue_transaction() to be sure nothing is added to the queue after this point?

File: src/osd/OSD.cc [hunk 6]
```diff
@@ -11128,6 +11203,25 @@ void OSD::ShardedOpWQ::_enqueue_front(OpSchedulerItem&& item)
...
+
+  for (; shard_index < osd->num_shards; shard_index++) {
+    auto& sdata = osd->shards[shard_index];
+    ceph_assert(sdata);
+    sdata->shard_lock.lock();
+    int work_count = 0;
+    while(! sdata->scheduler->empty() ) {
+      auto work_item = sdata->scheduler->dequeue();
+      work_count++;
+      // TBD - do we need to delete the work_item ???
```
Comment (jdurgin) [inline src/osd/OSD.cc:11219]: don't think it matters since we're about to exit

**Comment by jdurgin** [comment]
looking pretty good, the teuthology piece could be a separate PR

**Comment by benhanokh** [comment]
There are 2 tests showing corruption in the allocation file:
/a/yuriw-2022-02-15_22:35:42-rados-wip-yuri8-testing-2022-02-15-1214-distro-default-smithi/6685993/teuthology.log
/a/yuriw-2022-02-15_22:35:42-rados-wip-yuri8-testing-2022-02-15-1214-distro-default-smithi/6686090/teuthology.log

I need to understand what was done in those tests and why we ended with corrupted allocation file, but for now we should stop the merge :-(

**Comment by benhanokh** [comment]
It seems that the problem is an old race-condition in NCB code unrelated to safe-fast-shutdown.
The issue is detected by the extra sanity-checks/scrubs added in this PR.

The problem might be a race in the way we free up space on BlueFS on compaction

**Review by jdurgin — APPROVED** [review]
extra changes look good

**Comment by neha-ojha** [comment]
Being tested here: https://pulpito.ceph.com/nojha-2022-02-23_17:58:44-rados-GBH_safe_shutdown_v2_basecode_sanity_check_disabled_2-distro-basic-smithi/

**Comment by neha-ojha** [comment]
@benhanokh please see this failure https://pulpito.ceph.com/nojha-2022-02-23_17:58:44-rados-GBH_safe_shutdown_v2_basecode_sanity_check_disabled_2-distro-basic-smithi/6702922/

```
2022-02-23T20:48:36.298 INFO:teuthology.orchestra.run.smithi181.stderr:2022-02-23T20:48:36.317+0000 7f3525ca8700 -1 rocksdb: submit_common error: Corruption: block checksum mismatch: stored = 1277464416, computed = 3826455388  in db/000139.sst offset 1576903 size 4054 code = 2 Rocksdb transaction:
2022-02-23T20:48:36.299 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.OBJ_0000000000000003.1DB26E06.27b.smithi14140276-34..' value size = 74)
2022-02-23T20:48:36.299 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.SNA_3_0000000000000277_0000000000000003.1DB26E06.27b.smithi14140276-34..' value size = 70)
2022-02-23T20:48:36.299 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.OBJ_0000000000000003.1DB26E06.2ef.smithi14140276-34..' value size = 98)
2022-02-23T20:48:36.299 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.SNA_3_00000000000002DB_0000000000000003.1DB26E06.2ef.smithi14140276-34..' value size = 70)
2022-02-23T20:48:36.300 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.SNA_3_00000000000002DF_0000000000000003.1DB26E06.2ef.smithi14140276-34..' value size = 70)
2022-02-23T20:48:36.300 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.SNA_3_00000000000002E4_0000000000000003.1DB26E06.2ef.smithi14140276-34..' value size = 70)
2022-02-23T20:48:36.300 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.SNA_3_00000000000002ED_0000000000000003.1DB26E06.2ef.smithi14140276-34..' value size = 70)
2022-02-23T20:48:36.300 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.OBJ_0000000000000003.1DB26E06.31f.smithi14140276-34..' value size = 82)
2022-02-23T20:48:36.301 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.SNA_3_0000000000000316_0000000000000003.1DB26E06.31f.smithi14140276-34..' value size = 70)
2022-02-23T20:48:36.301 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.SNA_3_000000000000031C_0000000000000003.1DB26E06.31f.smithi14140276-34..' value size = 70)
2022-02-23T20:48:36.301 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.OBJ_0000000000000003.1DB26E06.33c.smithi14140276-34..' value size = 90)
2022-02-23T20:48:36.301 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.SNA_3_000000000000032E_0000000000000003.1DB26E06.33c.smithi14140276-34..' value size = 70)
2022-02-23T20:48:36.302 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.SNA_3_000000000000033B_0000000000000003.1DB26E06.33c.smithi14140276-34..' value size = 70)
2022-02-23T20:48:36.302 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x0000000000000000C03716250000000000000402'.SNA_3_000000000000033C_0000000000000003.1DB26E06.33c.smithi14140276-34..' value size = 70)
2022-02-23T20:48:36.302 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x00000000000000038BD46706000000000000'2(~' value size = 0)
2022-02-23T20:48:36.302 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = p key = 0x00000000000000038BD46706000000000000'2(-' value size = 0)
2022-02-23T20:48:36.302 INFO:teuthology.orchestra.run.smithi181.stderr:PutCF( prefix = O key = 0x7F80000000000000038BD46706'!smithi14140276-34!='0xFFFFFFFFFFFFFFFEFFFFFFFFFFFFFFFF6F value size = 1776)
2022-02-23T20:48:36.303 INFO:teuthology.orchestra.run.smithi181.stderr:MergeCF( prefix = T key = 0x0000000000000003 value size = 40)
2022-02-23T20:48:36.303 INFO:teuthology.orchestra.run.smithi181.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-10831-g40213c87/rpm/el8/BUILD/ceph-17.0.0-10831-g40213c87/src/os/bluestore/BlueStore.cc: In function 'void BlueStore::_txc_apply_kv(BlueStore::TransContext*, bool)' thread 7f3525ca8700 time 2022-02-23T20:48:36.318044+0000
2022-02-23T20:48:36.303 INFO:teuthology.orchestra.run.smithi181.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-10831-g40213c87/rpm/el8/BUILD/ceph-17.0.0-10831-g40213c87/src/os/bluestore/BlueStore.cc: 12773: FAILED ceph_assert(r == 0)
2022-02-23T20:48:36.304 INFO:teuthology.orchestra.run.smithi181.stderr: ceph version 17.0.0-10831-g40213c87 (40213c87bdc043acdadc03f47f4d21c98bee71d8) quincy (dev)
2022-02-23T20:48:36.304 INFO:teuthology.orchestra.run.smithi181.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x152) [0x7f3533a8d294]
2022-02-23T20:48:36.305 INFO:teuthology.orchestra.run.smithi181.stderr: 2: /usr/lib64/ceph/libceph-common.so.2(+0x2834b5) [0x7f3533a8d4b5]
2022-02-23T20:48:36.305 INFO:teuthology.orchestra.run.smithi181.stderr: 3: (BlueStore::_txc_apply_kv(BlueStore::TransContext*, bool)+0x5fb) [0x5611a0b42f9b]
2022-02-23T20:48:36.305 INFO:teuthology.orchestra.run.smithi181.stderr: 4: (BlueStore::_kv_sync_thread()+0x2139) [0x5611a0b8a429]
2022-02-23T20:48:36.305 INFO:teuthology.orchestra.run.smithi181.stderr: 5: (BlueStore::KVSyncThread::entry()+0x11) [0x5611a0bba4f1]
2022-02-23T20:48:36.305 INFO:teuthology.orchestra.run.smithi181.stderr: 6: /lib64/libpthread.so.0(+0x814a) [0x7f3532eb914a]
2022-02-23T20:48:36.306 INFO:teuthology.orchestra.run.smithi181.stderr: 7: clone()
```

File: src/os/bluestore/BlueStore.cc [hunk 7]
```diff
@@ -7681,12 +7688,15 @@ int BlueStore::umount()
...
+    // skip cache cleanup step on fast shutdown
+    if (likely(!m_fast_shutdown)) {
+      _shutdown_cache();
+    }
     dout(20) << __func__ << " closing" << dendl;
   }
-
   _close_db_and_around();
-  if (cct->_conf->bluestore_fsck_on_umount) {
+  // disable fsck on fast-shutdown
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:7698]: "skip" is better word then "disable" here

File: src/osd/OSD.cc [hunk 8]
```diff
@@ -4448,6 +4504,9 @@ int OSD::shutdown()
   hb_front_server_messenger->shutdown();
   hb_back_server_messenger->shutdown();
 
+  utime_t duration = ceph_clock_now() - start_time_func;
+  dout(0) <<"Slow Shutdown duration:" << duration << " seconds" << dendl;
```
Comment (aclamk) [inline src/osd/OSD.cc:4508]: Proposal: How about calling it "Full" or "Orderly" instead of "Slow"?

File: src/osd/OSD.cc [hunk 9]
```diff
@@ -4258,27 +4258,44 @@ PerfCounters* OSD::create_recoverystate_perf()
 
 int OSD::shutdown()
 {
+  // vstart overwrites osd_fast_shutdown value in the conf file -> force the value here!
+  //cct->_conf->osd_fast_shutdown = true;
+
+  dout(0) << "Fast Shutdown: - cct->_conf->osd_fast_shutdown = "
```
Comment (aclamk) [inline src/osd/OSD.cc:4264]: Need to improve output, show something like:
"Shutdown: Fast, null-fm=true"

File: src/osd/OSD.cc [hunk 10]
```diff
@@ -4258,27 +4258,44 @@ PerfCounters* OSD::create_recoverystate_perf()
 
 int OSD::shutdown()
 {
+  // vstart overwrites osd_fast_shutdown value in the conf file -> force the value here!
```
Comment (aclamk) [inline src/osd/OSD.cc:4261]: Suspicious comments. Is there something missing in this PR?
If this is for testing purposes, then it should be a separate commit.

File: src/osd/OSD.cc [hunk 11]
```diff
@@ -4258,27 +4258,44 @@ PerfCounters* OSD::create_recoverystate_perf()
...
+  if (!cct->_conf->osd_fast_shutdown) {
+    dout(0) << "shutdown" << dendl;
+  }
+
+  // don't accept new task for this OSD
   set_state(STATE_STOPPING);
 
-  // Debugging
-  if (cct->_conf.get_val<bool>("osd_debug_shutdown")) {
+  // Disabled debugging during fast-shutdown
```
Comment (aclamk) [inline src/osd/OSD.cc:4297]: We do not disable any debugging here....

**Review by aclamk — APPROVED** [review]
Looks good.

**Comment by ljflores** [comment]
http://pulpito.front.sepia.ceph.com/yuriw-2022-03-07_22:06:10-rados-wip-yuri-testing-2022-03-07-0958-distro-default-smithi/

Failures, unrelated:
    https://tracker.ceph.com/issues/50280
    https://tracker.ceph.com/issues/53680
    https://tracker.ceph.com/issues/51076

Details:
    1. cephadm: RuntimeError: uid/gid not found - Ceph
    2. ERROR:tasks.rook:'waiting for service removal' reached maximum tries (90) after waiting for 900 seconds - Ceph - Orchestrator
    3. "wait_for_recovery: failed before timeout expired" during thrashosd test with EC backend. - Ceph - RADOS

**Comment by ljflores** [comment]
jenkins test windows
